### ISSUE TITLE
HydrationBoundary not rendering page with updated cache, like Hydrate used to do.

### ISSUE THREAD

**Issue by emzoumpo** [TanStack/query#6240]
### Describe the bug

In React Query v5, <HydrationBoundary /> doesn't behave the same as <Hydrate /> in v4 (at least for Next.js pages router).
Despite the migration instructions, it doesn't seem to update the queryCache when the new page is rendering, using client side navigation.

It looks like the query cahce is updated too late, when the new page has already rendered.
In React Query v4 it works fine.

### Your minimal, reproducible example

Attached in Steps to reproduce

### Steps to reproduce

Minimal example with v5: 
https://codesandbox.io/p/sandbox/sharp-dhawan-qm6ykp?embed=1&file=%2Fsrc%2Fpages%2F%5B%5B...catchAll%5D%5D.js%3A27%2C42

Video of v5 (wrong):
https://github.com/TanStack/query/assets/2103443/6dcd1038-2844-47cc-97ae-be5833752ecf

-------

Minimal example with v4 (correct behaviour):
 https://codesandbox.io/p/sandbox/determined-fog-jjny9g?embed=1&file=%2Fsrc%2Fpages%2F%5B%5B...catchAll%5D%5D.js%3A27%2C40

Video of V4 (correct):
https://github.com/TanStack/query/assets/2103443/e580953a-a983-4ae4-810c-4916ef122430

 

### Expected behavior

As a user, I expect when using <HydrationBoundary />, the cache to be updated before the children render.

### How often does this bug happen?

Every time

### Platform

- Any platform

### Tanstack Query adapter

react-query

### TanStack Query version

v5

**Comment by mlazzarini** [comment]
@Ephem: is that possible that [this PR](https://github.com/TanStack/query/pull/5989) relates to the problem described in this issue? Thanks in advance for any helpful insight here.

**Comment by Ephem** [comment]
@emzoumpo Thanks for filing this issue.
@mlazzarini Thanks for doing some research and pinging me!

As that PR describes, and as is described [here](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#hydration-api-changes) in the migration guide, this is an intentional change.

In v4, a page transition could kick off, immediately put new data into the cache which would update the _current_ page unexpectedly even before the transition to the next page was done, and even if that transition was then aborted. We had an _observable_ side effect in render which is not allowed in React.

The new approach in v5 isn't perfect either. If the data already exists in the cache it is hydrated _after_ the initial render which can be jarring in some cases, but at least it is _eventually consistent_.

The story for having external caches like React Query work with React's concurrent features is a bit rough and while there might be ways to improve this in the future, in the meanwhile we want to opt for _correctness_.

You probably did it that way to showcase the problem, but just to be clear the reason the sandbox you provided isn't updating _at all_ is because `useQueryClient().getQueryData(["pageData"])` renders the current data in the cache but never sets up a observer to re-render when the data updates. Using a `useQuery` as in [this fork](https://codesandbox.io/p/sandbox/compassionate-sun-kc96nw?embed=1&file=%2Fsrc%2Fpages%2F%5B%5B...catchAll%5D%5D.js%3A18%2C10) works as expected.

I'm curious, did you file this issue because you have some current use case like the one in the sandbox where you are reading data from the cache directly, or because you noticed the page is rendering with old data first and then rerendering?

**Comment by emzoumpo** [comment]
Hi @ephem, thank you for your reply.

It's both. Of course the example is pretty simple to showcase the issue.

We use this pattern a lot, so we have hooks that read directly from cache without subscribing / using useQuery.

The reason for that is that there is a lot of data that is fetched / set on the server, when we call Next.js `getServerSideProps()`, which is called for both Server & Client Side navigation. 

This data **cannot and/or should not be fetched** on the client side, for security and/or performance reasons.
That means that the cache is populated with the relevant data inside of the `getServerSideProps()` with `prefetch()` and is already available to the page components, the same way `pageProps` are available to the page components when the new page begin rendering.

Looking at the linked PR, I think that the following statement is not true for Next.js Pages Router in general, since `<HydrationBoundary />` is called / used potentially on every page transition:

```
While this doesn't change the API in any way it's technically a breaking change, though I suspect and hope very few will notice any difference, especially since most people currently only use HydrationBoundary once during the initial page load and since all queries are new at that point, that behaviour should not have changed.
```

Also, I think it is very highly dependent on the app framework / scale / requirements and extend of use of React Query, whether the following statement makes it an ok tradeoff:

```
This means the new page will first render with existing data+new data, to immediately rerender to show only new data when the effect has run.
```

If the React Query cache contains a lot of data (that are being supposed to be replaced / merged when the new page renders), rerendering the page first with old date and then once again with new is something that can definitely have detrimental impact on the performance, responsiveness and layout shift of the app.

One of the benefits of Next.js is that you can prerender and / or prepopulate data so when the new page transition happens, data will be there in order to exactly avoid this kind of re-rendering after the page transition (when user expects data to be updated).

So, while I see the reason why this new behaviour is more correct, it comes with non-negligible tradeoffs.

In conclusion, I think that migrating to this new behaviour will be hard to do in our scale, since the options would be to fetch data on the client or add QueryCache listeners I suppose (please correct me if I'm wrong).
Right now this is the only thing that prevents us from upgrading from v4 to v5, and it doesn't look like it can be done incrementally.

If such behaviour is indeed desired for most users, would it be possible to have an option to opt-in to the old behaviour with a different Hydration component or property, so that we can have a chance of assessing and migrating our code gradually?

**Comment by TkDodo** [comment]
> We use this pattern a lot, so we have hooks that read directly from cache without subscribing / using useQuery.

reading from the cache during rendering without creating an observer was never a recommended pattern. What you'd want is to call useQuery and set enabled:false

**Comment by Ephem** [comment]
@emzoumpo Thanks for providing details around your use case! 

> I think that the following statement is not true for Next.js Pages Router in general

I agree, that formulation is a bit sloppy. I still hope few notice a huge difference though.

> If the React Query cache contains a lot of data (that are being supposed to be replaced / merged when the new page renders), rerendering the page first with old date and then once again with new is something that can definitely have detrimental impact on the performance, responsiveness and layout shift of the app.

> So, while I see the reason why this new behaviour is more correct, it comes with non-negligible tradeoffs.

This is all true. Another way to see it is in a way we are opting out of prerendering features when this happens. We still think favouring correctness is the correct path here though to avoid bugs and surprising behaviour. We fully acknowledge that React Query is not fully compatible with concurrent features.

As @TkDodo mentioned, reading in render is problematic and you want an observer. If you still want to mimic the old behaviour in v5 (even though it's not the correct tradeoff for a _library_, it might be right for you!), the low level `hydrate()` API is still available so you can build your own `<Hydrate>` that works as before. I think you can just copy-paste the code from v4 into it's own file and it should work.

(If anyone wants more context, there's an old, and slightly outdated, discussion about it in the React 18 working group: https://github.com/reactwg/react-18/discussions/121)

**Comment by emzoumpo** [comment]
Thank you very much for the detailed response, @Ephem!

I see now that they are cases in which the logic we're following might not be correct, especially as React, Next.js and React Query evolves, and new rendering modes are introduced.

I can understand that the new behaviour is more correct and an acceptable tradeoff for React Query, and we'd like to eventually migrate to properly support this.

That being said, there are a lot of nice features & improvements on React Query v5, that we wouldn't want to miss, until all of our code is refactored to account for the new behaviour.

I will try to include the code from v4, and hopefully that would be enough to migrate to v5 for now, and then gradually refactor our code so it is correct with the new behaviour, so we eventually remove the workaround.

Thank you once again for pointing out the possible solutions and offering a possible path to move forward!

**Comment by valeryq** [comment]
Hi @Ephem, I encountered a similar issue related to hydration queries. In this scenario, I have queries that are already in the cache but contain empty data.

For example, I have a page with both a `ClientComponent` and a `HydratedComponent`, and they both fetch the same query, as shown below:

```javascript
<>
  <ClientComponent />
  <HydratedComponent />
</>
```
As `ClientComponent` created query in cache, then `HydratedComponent` will not hydrate data in render and will put hydration process in a queue, as result I have empty data from cache for the first render, cause it wasn't hydrated.
Could we improve the condition in `HydrationBoundary` which checks whether query in cache or it has empty data?

I mean something like that:
```javascript
if (!existingQuery || !existingQuery.state.data) {
  newQueries.push(dehydratedQuery)
}
```

By making this change, the first render with empty data would have prefetched data, resulting in a more accurate representation of the content and we will use a hydration queue in case of react transitions to prevent showing new data until the transition is finished.
